tests/30rooms/13anonymousaccess.pl

```diff
@@ -120,15 +110,122 @@
 
                require_json_keys( $body, qw( chunk ) );
                $body->{chunk} >= 1 or die "Want at least one event";
-               my $event = $body->{chunk}[0];
+               my $event = $body->{chunk}->[0];
                require_json_keys( $event, qw( content ) );
                my $content = $event->{content};
                require_json_keys( $content, qw( body ) );
                $content->{body} eq "mice" or die "Want content body to be mice";
 
-               Future->done( 1 );
+               Future->done( $body->{end} );
             }),
-         );
+
+            matrix_send_room_text_message( $user, $room_id, body => "mice" )
+            ->on_done( sub {
+               ( $sent_event_id ) = @_;
+            }),
+         )->then( sub {
+            my ( $stream_token ) = @_;
+
+            Future->needs_all(
+               do_request_json_for( $anonymous_user,
+                  method => "GET",
+                  uri    => "/api/v1/events",
+                  params => {
+                     room_id => $room_id,
+                     from    => $stream_token,
+                  },
+               )->then( sub {
+                  my ( $body ) = @_;
+
+                  my $chunk = ignore_presence_for( [ $anonymous_user ], @{ $body->{chunk} } );
+                  log_if_fail "chunk", $chunk;
+                  @{ $chunk } == 1 or die "Want exactly one event";
+                  my $event = $chunk->[0];
+                  $event->{type} eq "m.presence" or die "Wrong event type";
+                  $event->{content}->{user_id} eq $user->user_id or die "Wrong user";
+
+                  Future->done( $body->{end} );
+               }),
+
+               Future->needs_all(
```

Comment:
This inner `needs_all` is also redundant since it's nested inside another one.
`all( ONE, all( TWO, THREE ) )` is the same as `all( ONE, TWO, THREE )`